src/Lecture4.hs

```diff
@@ -134,8 +143,43 @@ errors. We will simply return an optional result here.
 🕯 HINT: Use the 'readMaybe' function from the 'Text.Read' module.
 -}
 
+-- split :: Char -> String -> [String]
+-- split d string = helper string ""
+--   where
+--     helper [] acc = [acc]
+--     helper (x:xs) acc | d == x = acc : helper xs ""
+--                       | otherwise = helper xs (acc ++ [x])
+
+split :: Char -> String -> [String]
+split _ [] = [""]
+split d (x:xs) | x == d  = "" : split d xs
+               | otherwise = (x : y) : ys
+    where
+      (y:ys) = split d xs
+
+dropSpaces :: String -> String
+dropSpaces = takeWhile (not . isSpace) . dropWhile isSpace
+
+
+readTrade :: String -> Maybe TradeType
+readTrade "Buy" = Just Buy
+readTrade "Sell" = Just Sell
+readTrade _ = Nothing
+
+readMaybeNatural :: String -> Maybe Int
+readMaybeNatural str = do
+  n <- readMaybe str
+  guard $ n >= 0
+  pure n
+
 parseRow :: String -> Maybe Row
-parseRow = error "TODO"
+parseRow string = case split ',' string of
+                    ["", _, _] -> Nothing
+                    [prod, trade, cost] -> do
+                                           t <- readTrade . dropSpaces $ trade
+                                           c <- readMaybeNatural cost
+                                           pure (Row prod t c)
+                    _ -> Nothing
```

Comment:
You have an excellent monadic implementation 🔝 

I haven't explained the following during lectures but there's a nice way to write this parsing syntax using `do`-notation, `Monad` instance for `Maybe` and the `MonadFail` desugaring trick 🧙🏻 

You can read more about this trick here:

* https://kowainik.github.io/posts/haskell-mini-patterns#monadfail-sugar

Applying this programming pattern and a few helper functions, the `parseRow` function will look like this:

```suggestion
parseRow string = do
    [prod, trade, cost] <- Just $ split ',' string
    guard $ prod /= ""
    t <- readMaybe trade
    c <- readMaybeNatural cost
    pure $ Row prod t c
```